src/content/effective-dart/design.md

```diff
@@ -1678,11 +1667,13 @@ bool convertToBool(Object arg) {
 }
 ```
 
-The main exception to this rule is when working with existing APIs that use
-`dynamic`, especially inside a generic type. For example, JSON objects have type
-`Map<String, dynamic>` and your code will need to accept that same type. Even
-so, when using a value from one of these APIs, it's often a good idea to cast it
-to a more precise type before accessing members.
+Prefer using `Object?` over `dynamic` in code that is not invoking a member
+dynamically, even when working with existing APIs that use `dynamic`.
+For example, JSON objects have runtime type `Map<String, dynamic>`, which is an
```

Comment:
I'd say "JSON objects traditionally have the static type `Map<String, dynamic>`"

The runtime type doesn't matter, there is no meaningful difference between `dynamic` and `Object?` in a runtime type.

Also "JSON objects" is ambiguous. It can either mean any object created by `jsonDecode`, which is typed as just `dynamic`,  or it can mean a Dart representation of a "JSON Object" (based on a JavaScript Object), which is indeed traditionally represented by a map with static type `Map<String, dynamic>`.
The `dart: convert` code doesn't mandate that you use `Map<String, dynamic>` instead of `Map<String, Object?>` when you match or cast to a `Map` type, that's entirely user habit. It's this just saying that you should cast those to `... Object?...` before using them?